ql/src/java/org/apache/hadoop/hive/ql/Context.java

```diff
@@ -361,8 +361,16 @@ private DestClausePrefix getMergeDestClausePrefix(ASTNode curNode) {
     assert insert != null && insert.getType() == HiveParser.TOK_INSERT;
     ASTNode query = (ASTNode) insert.getParent();
     assert query != null && query.getType() == HiveParser.TOK_QUERY;
-
-    int tokFromIdx = query.getFirstChildWithType(HiveParser.TOK_FROM).getChildIndex();
+    ASTNode from = (ASTNode) query.getFirstChildWithType(HiveParser.TOK_FROM);
```

Comment:
According to the comment a few line above the code here assumes that the AST has a certain shape. I am not sure to what extend other parts of the code rely on this structure. Have you checked if there is a more intuitive fix when going from CBO to AST?

Reply:
I looked at several different places and thought that this was the place to fix it, but I am not very familiar with this so maybe I have missed something.

CBO plan is a HiveValues node for the failing query, and the AST looks fine for the plan. The query fails in `doPhase1`, so it might be possible to intercept it before:
```
case HiveParser.TOK_DESTINATION:
        ctx_1.dest = this.ctx.getDestNamePrefix(ast, qb).toString() + ctx_1.nextNum;
```
as it fails in `this.ctx.getDestNamePrefix(ast, qb)`. I will have to look into it.

Reply:
For the case of ACID tables do we have the same HiveValues CBO plan? If yes then maybe an option would be to check where ACID and Iceberg start to derive.

Reply:
If I remember correctly, for ACID tables the plan is different, but I will have to check again. I will look into it tomorrow.

Reply:
In case of ACID we always use `merge on read` and the merge statement is rewritten to a multi insert statement. In case of such statements only the from clause (query clause, the merge and using clause of the merge)  of the statement is optimized by CBO. This is usually a query with a join of the target and source tables. If the source is a query the plan can be more complex. However the predicate causing the always empty result is in the when matched clause of the merge (one of the insert branches of the multi insert) so it doesn't appear in the CBO plan and we can't end up with a `HiveValues(tuples=[[]])` CBO plan.

Reply:
As per the discussion, the logic for ACID tables differs significantly so we can't really use it here thus it is acceptable to retain the solution proposed here.